### PR TITLE
py/obj: Add support for __float__ converter function.

### PR DESCRIPTION
**Pull request by andrewleech**
I've run into a number of situation where I'm using a custom class that's intended to wrap an integer or float value, eg. https://github.com/micropython/micropython-lib/pull/502

It's confusing that other dunder converter functions like `__int__` and `__str__` work but not `__float__`. This PR resolves this, though in a different way to these other functions.

Another similar question was raised here: https://forum.micropython.org/viewtopic.php?f=2&t=6467

### CODE REVIEW

**Comment by andrewleech** [comment]
Ref: https://docs.python.org/3/reference/datamodel.html#object.__float__

**Comment by mattytrentini** [comment]
One point of discussion; if `__float__` is not present then Python will call `__index__` as a fallback (and expect an int which it will promote to a float). The behaviour is [neatly summarised here](https://blog.finxter.com/python-__float__-magic-method/). 

We looked at adding the conversion fallback behaviour to improve Python compatibility but found that it added enough complexity that we were questioning it's value. Further, the same behaviour (call `__index__` if the regular dunder conversion method isn't present) should also be applied to int and complex - but these were a little harder to implement.

I guess the question is should we bother? It seems like an esoteric corner of the language...

File: py/obj.c [hunk 1]
```diff
@@ -67,10 +67,10 @@ const mp_obj_type_t *MICROPY_WRAP_MP_OBJ_GET_TYPE(mp_obj_get_type)(mp_const_obj_
     } else if (mp_obj_is_obj(o_in)) {
         const mp_obj_base_t *o = MP_OBJ_TO_PTR(o_in);
         return o->type;
-    #if MICROPY_PY_BUILTINS_FLOAT
+        #if MICROPY_PY_BUILTINS_FLOAT
```
Comment (dpgeorge) [inline py/obj.c:70]: is it possible to not reformat these #'s ?

**Comment by dpgeorge** [comment]
> It's confusing that other dunder converter functions like __int__ and __str__ work but not __float__. This PR resolves this, though in a different way to these other functions.

Did you consider using the same approach as those existing ones, using `MP_UNARY_OP_FLOAT`?

**Comment by andrewleech** [comment]
> Did you consider using the same approach as those existing ones, using `MP_UNARY_OP_FLOAT`?

I briefly did but saw the comment here about if forcing a bytecode update: https://github.com/micropython/micropython/blob/07cae9178fd409e8269d55d0b614612a7db49bef/py/runtime0.h#L66

I presume this was is more efficient? I haven't followed that code path enough to know how it works yet.

File: py/obj.c (hunk 1 above)
Comment (andrewleech) [reply py/obj.c:70]: Ah whoops, the first ~5 times I pushed this branch I undid all the auto-codeformat on these lines, must have missed that on the current / last push. Not sure why my uncrustify wants to do this here :-(

**Comment by andrewleech** [comment]
Looking closer at the existing `__int__` UNARY handling:
```

**Comment by dpgeorge** [comment]
> Is the point of all these functions that they return a small int / non-gc object?

No.  It's just that the `mp_unary_op_method_name` array has byte-size elements, not qstr-size elements.

**Comment by andrewleech** [comment]
@dpgeorge new approach added in https://github.com/micropython/micropython/pull/8899

**Comment by dpgeorge** [comment]
Superseded by #8899.

**Pull request closed without merging**